src/cpp/src/storage/checkpointer.cpp

```diff
@@ -15,10 +18,35 @@ Checkpointer::Checkpointer(std::shared_ptr<Model> model, shared_ptr<GraphModelSt
     config_ = config;
 }
 
-void Checkpointer::save(string checkpoint_dir, CheckpointMeta checkpoint_meta) {
+void Checkpointer::create_checkpoint(string checkpoint_dir, CheckpointMeta checkpoint_meta, int epochs) {
+    string tmp_checkpoint_dir = checkpoint_dir + "checkpoint_" + std::to_string(epochs) + "_tmp/";
+    std::filesystem::create_directory(tmp_checkpoint_dir);
 
-    if (checkpoint_meta.has_model) {
+    std::string new_embeddings_file = tmp_checkpoint_dir + PathConstants::embeddings_file + PathConstants::file_ext;
+    std::string new_embeddings_state_file = tmp_checkpoint_dir + PathConstants::embeddings_state_file + PathConstants::file_ext;
+
+    std::string embeddings_file = checkpoint_dir + PathConstants::embeddings_file + PathConstants::file_ext;
+    std::string embeddings_state_file = checkpoint_dir + PathConstants::embeddings_state_file + PathConstants::file_ext;
+
+    if (std::filesystem::exists(embeddings_file)) {
+        if (this->config_->save_encoded) std::filesystem::copy(embeddings_file, new_embeddings_file);
```

Comment:
The save_encoded option can be removed and the embeddings can be copied always.

The original idea behind the save_encoded option was to save the output of the last layer of the encoder at each checkpoint. But it's fine to not include that functionality.